File: src/adapters/claude-code.ts
import type {
  HarnessAdapter,
  HookEntry,
  PortableHooks,
} from "../domain/harness.js";
import {
  buildBaseAgentArtifact,
  buildBaseManifest,
  pascalMatcherHooks,
} from "./_shared.js";

const CLAUDE_AGENT_KEYS: ReadonlySet<string> = new Set([
  "skills",
  "color",
  "effort",
  "disallowedTools",
  "permissionMode",
]);

const CLAUDE_MANIFEST_KEYS = [
  // "agents" intentionally omitted — Claude Code's plugin manifest validator
  // rejects this key (see issue #58). Files under the bundle's default
  // `./agents/` directory are auto-discovered at startup, so the conventional
  // layout cheese-flow uses still works without it. Re-add when upstream
  // accepts the field again.
  "skills",
  "commands",
  "hooks",
  "mcpServers",
] as const;

// Hook command paths must resolve from the installed plugin root in Claude
// Code, not the user's working directory. The portable `hooks.json` source
// uses `bash hooks/cheese-bootstrap.sh`; rewrite that to reference
// `${CLAUDE_PLUGIN_ROOT}` so the script is found wherever Claude Code installs
// the plugin. Other commands pass through unchanged.
const BOOTSTRAP_RELATIVE_PATH = "hooks/cheese-bootstrap.sh";

function rewriteBootstrapCommand(entry: HookEntry): HookEntry {
  if (!entry.command.includes(BOOTSTRAP_RELATIVE_PATH)) return entry;
  return {
    ...entry,
    command: entry.command.replace(
      BOOTSTRAP_RELATIVE_PATH,
      `\${CLAUDE_PLUGIN_ROOT}/${BOOTSTRAP_RELATIVE_PATH}`,
    ),
  };
}

function rewriteBootstrapPaths(portable: PortableHooks): PortableHooks {
  const result: PortableHooks = {};
  for (const [event, entries] of Object.entries(portable)) {
    if (entries === undefined) continue;
    result[event as keyof PortableHooks] = entries.map(rewriteBootstrapCommand);
  }
  return result;
}

export const claudeCodeAdapter: HarnessAdapter = {
  name: "claude-code",
  displayName: "Claude Code",
  outputRoot: ".claude",
  agentDirectory: "agents",
  skillDirectory: "skills",
  commandDirectory: "commands",
  defaultModel: "sonnet",
  notes: [
    "Use concise markdown headings and explicit tool guidance.",
    "Prefer Claude model identifiers in agent metadata and output.",
  ],
  manifestDir: ".claude-plugin",
  buildManifest: (metadata, componentPaths) =>
    buildBaseManifest(metadata, componentPaths, CLAUDE_MANIFEST_KEYS),
  mcpFileName: ".mcp.json",
  buildHookConfig: (portable) => ({
    hooks: pascalMatcherHooks(rewriteBootstrapPaths(portable)),
  }),
  buildAgentArtifact: (input) =>
    buildBaseAgentArtifact(input, CLAUDE_AGENT_KEYS),
  capabilities: {
    skillFrontmatterKeys: new Set(["model", "context"]),
    agentFrontmatterKeys: CLAUDE_AGENT_KEYS,
    hookEvents: new Set([
      "sessionStart",
      "sessionEnd",
      "preToolUse",
      "postToolUse",
      "stop",
      "subagentStop",
      "notification",
      "preCompact",
      "userPromptSubmit",
    ]),
    toolNames: new Set([
      "Agent",
      "Task",
      "NotebookEdit",
      "WebSearch",
      "WebFetch",
      "TodoWrite",
    ]),
    bootstrapHook: true,
  },
};

File: python/cheese_flow/adapters/claude_code.py
"""Port of `src/adapters/claude-code.ts`."""

from __future__ import annotations

from typing import Any

from cheese_flow.adapters._shared import (
    build_base_agent_artifact,
    build_base_manifest,
    camel_case_hooks,
)
from cheese_flow.lib.harness import (
    AgentArtifact,
    AgentArtifactInput,
    HarnessAdapter,
    HarnessCapabilities,
    ManifestComponentPaths,
    PluginMetadata,
    PortableHooks,
)

CLAUDE_AGENT_KEY_ORDER: tuple[str, ...] = (
    "skills",
    "color",
    "effort",
    "disallowedTools",
    "permissionMode",
)
CLAUDE_AGENT_KEYS: frozenset[str] = frozenset(CLAUDE_AGENT_KEY_ORDER)

CLAUDE_MANIFEST_KEYS: tuple[str, ...] = (
    "agents",
    "skills",
    "commands",
    "hooks",
    "mcpServers",
)


def _build_manifest(
    metadata: PluginMetadata,
    component_paths: ManifestComponentPaths,
) -> dict[str, Any]:
    return build_base_manifest(metadata, component_paths, CLAUDE_MANIFEST_KEYS)


def _build_hook_config(portable: PortableHooks) -> dict[str, Any]:
    return {"hooks": camel_case_hooks(portable)}


def _build_agent_artifact(artifact_input: AgentArtifactInput) -> AgentArtifact:
    return build_base_agent_artifact(artifact_input, CLAUDE_AGENT_KEY_ORDER)


claude_code_adapter = HarnessAdapter(
    name="claude-code",
    displayName="Claude Code",
    outputRoot=".claude",
    agentDirectory="agents",
    skillDirectory="skills",
    commandDirectory="commands",
    defaultModel="sonnet",
    notes=(
        "Use concise markdown headings and explicit tool guidance.",
        "Prefer Claude model identifiers in agent metadata and output.",
    ),
    manifestDir=".claude-plugin",
    buildManifest=_build_manifest,
    mcpFileName=".mcp.json",
    buildHookConfig=_build_hook_config,
    buildAgentArtifact=_build_agent_artifact,
    capabilities=HarnessCapabilities(
        skillFrontmatterKeys=frozenset({"model", "context"}),
        agentFrontmatterKeys=CLAUDE_AGENT_KEYS,
        hookEvents=frozenset(
            {
                "sessionStart",
                "sessionEnd",
                "preToolUse",
                "postToolUse",
                "stop",
                "subagentStop",
                "notification",
                "preCompact",
                "userPromptSubmit",
            },
        ),
        toolNames=frozenset(
            {
                "Agent",
                "Task",
                "NotebookEdit",
                "WebSearch",
                "WebFetch",
                "TodoWrite",
            },
        ),
        bootstrapHook=True,
    ),
)
